src/effects/stat_change.c

```diff
@@ -1,28 +1,230 @@
+    EffectBlueprint bp;
+    EffectBlueprint* bpPtr = &bp;
+    EffectInstance* effect;
+    StatChangeFXData* part;
+    s32 numParts = 1;
+
+    bp.init = stat_change_init;
+    bp.update = stat_change_update;
+    bp.renderWorld = stat_change_render;
+    bp.unk_14 = func_E00AC288;
+    bp.unk_00 = 0;
+    bp.effectID = EFFECT_STAT_CHANGE;
+
+    effect = shim_create_effect_instance(bpPtr);
+    effect->numParts = numParts;
+    part = effect->data = shim_general_heap_malloc(numParts * sizeof(*part));
+
+    ASSERT(effect->data != NULL);
 
-s32 D_E00AC83C[] = { 0xFF96FF82, 0xFF64FF46, 0xFF3CFF50, 0xFF6EFF73, 0xFF6BFF5C, 0xFF5AFF5E, 0xFF67FF68, 0xFF67FF64, 0xFF62FF63, 0xFF65FF64, 0xF064DC64, 0xC864B464, 0xA0648264, 0x64644664, 0x32642864, 0x1E641464, 0x0A640000 };
+    part->unk_00 = arg0;
+    part->pos.x = x;
+    part->pos.y = y;
+    part->pos.z = z;
+    part->timeLeft = time;
 
-s32 D_E00AC880[] = { 0x00000000, 0x00000001, 0x00000000, 0x00000002, 0x00000000, 0x00000003, 0x00000001, 0xFFFFFFFF, 0x00000001, 0xFFFFFFFE, 0x00000001, 0xFFFFFFFD, 0x00000002, 0x00000001, 0x00000002, 0x00000002, 0x00000002, 0x00000003, 0x00000000, 0x00000000 };
+    part->unk_18 = 0;
+    part->scale = scale;
+    part->unk_24 = 255;
+    part->scaleX = 0;
+    part->scaleY = 0;
+    part->unk_30 = 0;
+    part->unk_38 = 0;
+    part->unk_34 = 0;
 
-INCLUDE_ASM(s32, "effects/stat_change", stat_change_main);
+    part->arrowType = D_E00AC880[arg0].type;
+    part->arrowValue = D_E00AC880[arg0].value;
 
-void stat_change_init(void) {
+    return effect;
 }
 
-INCLUDE_ASM(s32, "effects/stat_change", stat_change_update);
+void stat_change_init(EffectInstance* effect) {
+}
+
+void stat_change_update(EffectInstance* effect) {
+    s32 old_unk18;
+    StatChangeFXData* part = effect->data;
+
+    part->timeLeft--;
+    old_unk18 = part->unk_18++;
+
+    if (part->timeLeft < 0) {
+        shim_remove_effect(effect);
+        return;
+    }
+
+    if (part->arrowType == ARROW_TYPE_ATK_DOWN) {
+        if (old_unk18 >= ARRAY_COUNT(D_E00AC83C)) {
+            part->unk_24 = 0;
+            part->scaleX = 0.0f;
+            part->scaleY = 0.0f;
+        } else {
+            part->unk_24 = D_E00AC83C[old_unk18].unk_00;
+            part->scaleY = (f32)D_E00AC83C[old_unk18].unk_01 * 0.01;
+            part->scaleX = 1.0f - part->scaleY + 1.0f;
+        }
+    } else {
+        if (old_unk18 >= ARRAY_COUNT(D_E00AC7F8)) {
+            part->unk_24 = 0;
+            part->scaleX = 0.0f;
+            part->scaleY = 0.0f;
+        } else {
+            part->unk_24 = D_E00AC7F8[old_unk18].unk_00;
+            part->scaleX = (f32)D_E00AC7F8[old_unk18].unk_01 * 0.01;
+            part->scaleY = 1.0f - part->scaleX + 1.0f;
+        }
+    }
 
-void stat_change_render(void) {
+    part->unk_3C = 0;
+    part->unk_40 = 0;
+
+    do {} while (0); // required to match
+
+    part->pos.x += part->unk_30;
+    part->pos.y += part->unk_38;
+    part->pos.z += part->unk_34;
```

Comment:
You can probably make unk_30 a Vec3f

Reply:
well it would be weird that they do `pos.y += unk_30.z` and `pos.z += part->unk_34.y`

Reply:
oh, my bad lol